From 9dcb3904458105f17b20268253c6d603dc7f9257 Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Thu, 4 Sep 2025 20:04:25 +0200
Subject: [PATCH 180/553] QWidget: fix propagating style with descendant
 selectors

The fix for QTBUG-133332 introduced a check to avoid an endless
recursion within setStyle_helper() but it prevents the correct
propagation of the style with a descendant selector. Therefore use
another approach and make sure QWidgetPrivate::inheritStyle() is not
called recursivly.
This amends 3252e1808c12c21f27bb4844a1497d18587a64b5.

Pick-to: 6.9
Task-number: QTBUG-133332
Fixes: QTBUG-139924
Change-Id: Ia0a1eec652380397f861364bbdc303dfd17b34f3
Reviewed-by: Richard Moe Gustavsen <richard.gustavsen@qt.io>
(cherry picked from commit 14aa698cc310dbe77b94752d80054fc4086d5c26)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/widgets/kernel/qwidget.cpp                |  9 ++++-
 src/widgets/kernel/qwidget_p.h                |  1 +
 .../widgets/kernel/qwidget/tst_qwidget.cpp    | 38 +++++++++++++++++--
 3 files changed, 44 insertions(+), 4 deletions(-)

diff --git a/src/widgets/kernel/qwidget.cpp b/src/widgets/kernel/qwidget.cpp
index d3d13649c94..3aeb77c1b53 100644
--- a/src/widgets/kernel/qwidget.cpp
+++ b/src/widgets/kernel/qwidget.cpp
@@ -160,6 +160,7 @@ QWidgetPrivate::QWidgetPrivate(decltype(QObjectPrivateVersion) version)
       , childrenHiddenByWState(0)
       , childrenShownByExpose(0)
       , dontSetExplicitShowHide(0)
+      , inheritStyleRecursionGuard(0)
 #if defined(Q_OS_WIN)
       , noPaintOnScreen(0)
 #endif
@@ -2680,7 +2681,7 @@ void QWidgetPrivate::setStyle_helper(QStyle *newStyle, bool propagate)
     extra->style = newStyle;
 
     // repolish
-    if (polished && q->windowType() != Qt::Desktop && oldStyle != q->style()) {
+    if (polished && q->windowType() != Qt::Desktop) {
         oldStyle->unpolish(q);
         q->style()->polish(q);
     }
@@ -2728,6 +2729,12 @@ void QWidgetPrivate::inheritStyle()
         proxy->repolish(q);
         return;
     }
+    if (inheritStyleRecursionGuard)
+        return;
+    inheritStyleRecursionGuard = true;
+    const auto resetGuard = qScopeGuard([&]() {
+            inheritStyleRecursionGuard = false;
+        });
 
     QStyle *origStyle = proxy ? proxy->base : extraStyle;
     QWidget *parent = q->parentWidget();
diff --git a/src/widgets/kernel/qwidget_p.h b/src/widgets/kernel/qwidget_p.h
index 991b932ecb8..981f8550f61 100644
--- a/src/widgets/kernel/qwidget_p.h
+++ b/src/widgets/kernel/qwidget_p.h
@@ -735,6 +735,7 @@ public:
     uint childrenHiddenByWState : 1;
     uint childrenShownByExpose : 1;
     uint dontSetExplicitShowHide : 1;
+    uint inheritStyleRecursionGuard : 1;
 
     // *************************** Focus abstraction ************************************
     enum class FocusDirection {
diff --git a/tests/auto/widgets/kernel/qwidget/tst_qwidget.cpp b/tests/auto/widgets/kernel/qwidget/tst_qwidget.cpp
index dcde8c19c75..0d31ff77a73 100644
--- a/tests/auto/widgets/kernel/qwidget/tst_qwidget.cpp
+++ b/tests/auto/widgets/kernel/qwidget/tst_qwidget.cpp
@@ -11,7 +11,9 @@
 #include <qlabel.h>
 #include <qlayout.h>
 #include <qlineedit.h>
-#include <qlistview.h>
+#if QT_CONFIG(listwidget)
+#include <qlistwidget.h>
+#endif
 #include <qmessagebox.h>
 #include <qmimedata.h>
 #include <qpainter.h>
@@ -22,6 +24,7 @@
 #include <qstylefactory.h>
 #include <private/qwidget_p.h>
 #include <private/qwidgetrepaintmanager_p.h>
+#include <private/qwindowsstyle_p.h>
 #include <private/qapplication_p.h>
 #include <private/qhighdpiscaling_p.h>
 #include <qcalendarwidget.h>
@@ -6806,7 +6809,32 @@ void tst_QWidget::deleteStyle()
     QCoreApplication::processEvents();
 }
 
-class TestStyle : public QCommonStyle
+
+#if QT_CONFIG(listwidget)
+class DontCrashOnSetStyleWidget : public QWidget
+{
+    Q_OBJECT
+public:
+    DontCrashOnSetStyleWidget()
+    {
+        lw = new QListWidget;
+        lwi = new QListWidgetItem;
+        lw->addItem(lwi);
+        lw->setItemWidget(lwi, new QLabel(u"test"_s));
+        auto l = new QVBoxLayout(this);
+        l->addWidget(lw);
+    }
+    bool testStyleSheetTarget() const
+    {
+        return lw->itemWidget(lwi)->testAttribute(Qt::WA_StyleSheetTarget);
+    }
+private:
+    QListWidget *lw = nullptr;
+    QListWidgetItem *lwi = nullptr;
+};
+#endif
+
+class TestStyle : public QWindowsStyle
 {
     void polish(QWidget *w) override
     {
@@ -6827,13 +6855,17 @@ void tst_QWidget::dontCrashOnSetStyle()
     });
     {
         qApp->setStyle(new TestStyle);
-        qApp->setStyleSheet("blub");
+        qApp->setStyleSheet(u"DontCrashOnSetStyleWidget QLabel {color:red;}"_s);
         QComboBox w;
         w.show();
         QVERIFY(QTest::qWaitForWindowExposed(&w));
         // this created an infinite loop / stack overflow inside setStyle_helper()
         // directly call polish instead waiting for the polish event
         qApp->style()->polish(&w);
+#if QT_CONFIG(listwidget)
+        DontCrashOnSetStyleWidget widget;
+        QVERIFY(widget.testStyleSheetTarget());
+#endif
     }
 }
 
-- 
2.51.2

